File: intellij-lsp/src/com/github/gtache/lsp/settings/server/LSPConfiguration.scala
package com.github.gtache.lsp.settings.server

import java.io.File

import com.github.gtache.lsp.settings.server.parser.ConfigurationParser

import scala.collection.JavaConverters._

case class LSPConfiguration(settings: Map[String, Map[String, AnyRef]]) {
  def getJavaSettings: java.util.Map[String, java.util.Map[String, AnyRef]] = {
    settings.map(pair => pair._1 -> pair._2.asJava).asJava
  }

  def getSettings: Map[String, Map[String, AnyRef]] = settings

  def isValid: Boolean = this != LSPConfiguration.invalidConfiguration

  def getScopeForUri(uri: String): String = "global" //TODO manage different scopes

  def getAttributesForSectionAndScope(section: String, scope: String = "global"): Map[String, AnyRef] = {
    settings.get(scope).map(m => m.filterKeys(key => key.startsWith(section)).map(keyValue => {
      val strippedKey = keyValue._1.stripPrefix(section)
      if (strippedKey.startsWith(".")) {
        strippedKey.stripPrefix(".") -> keyValue._2
      } else {
        strippedKey -> keyValue._2
      }
    })).orNull
  }

  def getAttributesForSectionAndUri(section: String, uri: String): Map[String, AnyRef] = {
    getAttributesForSectionAndScope(section, getScopeForUri(uri))
  }
}

object LSPConfiguration {
  def fromFile(file: File): LSPConfiguration = {
    ConfigurationParser.getConfiguration(file)
  }

  //TODO manage User config < Workspace config
  def fromFiles(files: Seq[File]): LSPConfiguration = {
    val configs = files.map(fromFile).map(_.settings)
    LSPConfiguration(configs.foldLeft(Map[String, Map[String, AnyRef]]())((configTop, configBottom) => {
      ConfigurationParser.combineConfigurations(configTop, configBottom)
    }))
  }

  val emptyConfiguration: LSPConfiguration = LSPConfiguration(Map("global" -> Map[String, AnyRef]()))
  val invalidConfiguration: LSPConfiguration = LSPConfiguration(null)

}

File: intellij-lsp/src/main/kotlin/com/github/gtache/lsp/settings/server/LSPConfiguration.kt
package com.github.gtache.lsp.settings.server

import com.github.gtache.lsp.settings.server.parser.ConfigurationParser
import java.io.File

data class LSPConfiguration(val settings: Map<String, Map<String, Any?>>) {

    companion object {
        fun fromFile(file: File): LSPConfiguration? {
            return ConfigurationParser.getConfiguration(file)
        }

        //TODO manage User config < Workspace config
        fun fromFiles(files: List<File>): LSPConfiguration {
            val configs = files.map(this::fromFile).mapNotNull { f -> f?.settings }
            return LSPConfiguration(configs.fold(HashMap()) { configTop: Map<String, Map<String, Any?>>, configBottom ->
                ConfigurationParser.combineConfigurations(configTop, configBottom)
            })
        }

        val emptyConfiguration: LSPConfiguration = LSPConfiguration(mapOf(Pair("global", HashMap())))
        val invalidConfiguration: LSPConfiguration = LSPConfiguration(emptyMap())
    }

    fun isValid(): Boolean = this !== invalidConfiguration

    fun getScopeForUri(uri: String): String = "global" //TODO manage different scopes

    fun getAttributesForSectionAndScope(section: String, scope: String = "global"): Map<String, Any?>? {
        return if (isValid()) {
            settings[scope]?.filter { m -> m.key.startsWith(section) }?.map { e ->
                val strippedKey = e.key.removePrefix(section)
                if (strippedKey.startsWith(".")) {
                    Pair(strippedKey.removePrefix("."), e.value)
                } else {
                    Pair(strippedKey, e.value)
                }
            }?.toMap()
        } else null
    }


    fun getAttributesForSectionAndUri(section: String, uri: String): Map<String, Any?>? {
        return getAttributesForSectionAndScope(section, getScopeForUri(uri))
    }
}